tracer/src/Datadog.Trace/AppSec/Waf/Context.cs

```diff
@@ -16,48 +18,66 @@ namespace Datadog.Trace.AppSec.Waf
     internal class Context : IContext
     {
         private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<Context>();
-        private readonly IntPtr contextHandle;
-        private readonly WafNative wafNative;
-        private readonly Encoder encoder;
-        private readonly List<Obj> argCache = new();
+        private readonly IntPtr _contextHandle;
+        private readonly Waf _waf;
+        private readonly ReaderWriterLock _wafLocker;
+        private readonly List<Obj> _argCache = new();
         private readonly Stopwatch _stopwatch;
-        private bool disposed;
+
+        // this lock is for locking the context handle and argCache objects, which aren't thread safe
+        private bool _disposed;
         private ulong _totalRuntimeOverRuns;
 
-        public Context(IntPtr contextHandle, WafNative wafNative, Encoder encoder)
+        // Beware this class is created on a thread but can be disposed on another so don't trust the lock is not going to be held
+        private Context(IntPtr contextHandle, Waf waf, ReaderWriterLock wafLocker, out bool valid)
         {
-            this.contextHandle = contextHandle;
-            this.wafNative = wafNative;
-            this.encoder = encoder;
+            _wafLocker = wafLocker;
+            // in high concurrency, the waf passed as argument here could have been disposed just above in betwen creation / waf update so last test here
+            valid = (_wafLocker.IsReadLockHeld || _wafLocker.EnterReadLock()) && !waf.Disposed;
```

Comment:
wouldn't it be better to check this before creating the object? It would save the allocation and using an out parameter seems slightly awkward.

Reply:
it was to save a few lines from locking but yes it s weird, fixed in 9daf1e931a478a25f20f4c693fc374289724d86d